hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java

```diff
@@ -0,0 +1,218 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.client;
+
+import com.hazelcast.client.impl.ClientDelegatingFuture;
+import com.hazelcast.client.impl.clientside.ClientMessageDecoder;
+import com.hazelcast.client.impl.clientside.HazelcastClientInstanceImpl;
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.client.impl.protocol.codec.SqlCloseCodec;
+import com.hazelcast.client.impl.protocol.codec.SqlExecuteCodec;
+import com.hazelcast.client.impl.protocol.codec.SqlFetchCodec;
+import com.hazelcast.client.impl.spi.impl.ClientInvocation;
+import com.hazelcast.client.impl.spi.impl.ClientInvocationFuture;
+import com.hazelcast.internal.nio.Connection;
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.internal.serialization.InternalSerializationService;
+import com.hazelcast.sql.SqlErrorCode;
+import com.hazelcast.sql.SqlException;
+import com.hazelcast.sql.SqlQuery;
+import com.hazelcast.sql.SqlResult;
+import com.hazelcast.sql.SqlService;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.QueryId;
+import com.hazelcast.sql.impl.QueryUtils;
+import com.hazelcast.sql.impl.row.Row;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Client-side implementation of SQL service.
+ */
+public class SqlClientService implements SqlService {
+
+    private static final ClientMessageDecoder<SqlExecuteResponse> EXECUTE_DECODER = clientMessage -> {
+        SqlExecuteCodec.ResponseParameters response = SqlExecuteCodec.decodeResponse(clientMessage);
+
+        return new SqlExecuteResponse(response.queryId, response.rowMetadata, response.rowPage, response.error);
+    };
+
+    private static final ClientMessageDecoder<SqlFetchResponse> FETCH_DECODER = clientMessage -> {
+        SqlFetchCodec.ResponseParameters response = SqlFetchCodec.decodeResponse(clientMessage);
+
+        return new SqlFetchResponse(response.rowPage, response.error);
+    };
+
+    private static final ClientMessageDecoder<Void> CLOSE_DECODER = clientMessage -> {
+        SqlCloseCodec.decodeResponse(clientMessage);
+
+        return null;
+    };
+
+    private final HazelcastClientInstanceImpl client;
+
+    public SqlClientService(HazelcastClientInstanceImpl client) {
+        this.client = client;
+    }
+
+    @Nonnull
+    @Override
+    public SqlResult query(@Nonnull SqlQuery query) {
+        Connection connection = client.getConnectionManager().getRandomConnection(true);
+
+        if (connection == null) {
+            throw rethrow(QueryException.error(
+                SqlErrorCode.CONNECTION_PROBLEM,
+                "Client must be connected to at least one data member to execute SQL queries"
```

Comment:
So a unısocket client that is connected to a lite member can not run a query? 
This is not suited to our unisocket client behavior. Our unisocket clients should be able to operate same regardless of the member type they are connected to.

Reply:
You can check `CreateProxyMessageTask`. It is one of the places that we explicitly redirect to a data member. 

The partition-based ones solve the problem automatically since they will run on their partition owner anyway. Even if we invoke the operation on a lite member, the invocation system redirects it to the partition owner.

Reply:
Yes, SQL may not work if there lite members in the topology. This is ok for the beta stage at least. Lite member support might be added in the future if we see a demand from users. 
This limitation will be mentioned explicitly in the product's documentation.

Reply:
I think it's pretty bad that it will fail to work randomly, depending on the member the non-smart client happens to connect. Users might try it and see that it works and then have this misleading error at runtime. In that case we should disallow non-smart clients entirely.

Reply:
Documentation solves this. Again - we do not support lite members not because we are lazy, or because we do not understand the implications. This is a necessary measure given the huge scope. 
Moreover, in the current PR, we iterate over all members to find the data member for SQL. The flakiness you mentioned is only possible if a user will flip-flop member state from lite member and vice versa. This is extremely unlikely.

Reply:
Iterating all connections does not help unisocket client. It will only have one connection. If the client disconnect from data member and connects to lite member, then it will fail. 

If this is the decision, we basically say that for the uniscoket client, if you use query, your cluster should not contain a lite member. For the smart client, I agree that the problem is very unlikely to occur.